cdc/cdc.cc

```diff
@@ -64,12 +65,119 @@ using namespace std::chrono_literals;
 
 static logging::logger cdc_log("cdc");
 
-class cdc::cdc_service::impl {
+namespace cdc {
+static schema_ptr create_log_schema(const schema&, std::optional<utils::UUID> = {});
+static schema_ptr create_stream_description_table_schema(const schema&, std::optional<utils::UUID> = {});
+static future<> populate_desc(db_context ctx, const schema& s);
+}
+
+class cdc::cdc_service::impl : service::migration_listener::empty_listener {
     db_context _ctxt;
 public:
     impl(db_context ctxt)
         : _ctxt(std::move(ctxt))
-    {}
+    {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.register_listener(this);
+        }
+    }
+    ~impl() {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.unregister_listener(this);
+        }
+    }
+
+    void on_before_create_column_family(const schema& schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        if (schema.cdc_options().enabled()) {
+            auto& db = _ctxt._proxy.get_db().local();
+            auto logname = log_name(schema.cf_name());
+            if (!db.has_schema(schema.ks_name(), logname)) {
+                // in seastar thread
+                auto log_schema = create_log_schema(schema);
+                auto stream_desc_schema = create_stream_description_table_schema(schema);
+                auto& keyspace = db.find_keyspace(schema.ks_name());
+
+                auto log_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), log_schema, timestamp);
+                auto stream_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), stream_desc_schema, timestamp);
+
+                mutations.insert(mutations.end(), std::make_move_iterator(log_mut.begin()), std::make_move_iterator(log_mut.end()));
+                mutations.insert(mutations.end(), std::make_move_iterator(stream_mut.begin()), std::make_move_iterator(stream_mut.end()));
+            }
+        }
+    }
+
+    void on_before_update_column_family(const schema& new_schema, const schema& old_schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        bool is_cdc = new_schema.cdc_options().enabled();
+        bool was_cdc = old_schema.cdc_options().enabled();
+
+        if (is_cdc != was_cdc || is_cdc) {
+            auto logname = log_name(old_schema.cf_name());
+            auto descname = desc_name(old_schema.cf_name());
+            auto& db = _ctxt._proxy.get_db().local();
+            auto& keyspace = db.find_keyspace(old_schema.ks_name());
+            auto log_schema = was_cdc ? db.find_column_family(old_schema.ks_name(), logname).schema() : nullptr;
+            auto stream_desc_schema = was_cdc ? db.find_column_family(old_schema.ks_name(), descname).schema() : nullptr;
+
+            assert(!was_cdc || log_schema);
```

Comment:
These asserts are rather pointless, considering you defined log_schema just two lines above, to be non-null exactly when was_cdc?

Reply:
It was my suggestion, but now you're saying:

> If the find_column_family() above failed, we would get an exception here, not a null.

So my suggestion was stupid indeed.

Reply:
pointless, yes, but otoh makes sure noone messes with the assignment. Otoh, I really wish our asserts were debug-only so one could drizzle them more abundantly.

Reply:
Note that asserts cause crashes of the entire node - and if they get caught on multiple nodes - the entire cluster. If we have a bug that effects just a specific operation and not the overall data integrity, it's better to throw an exception and not assert().
And, since in this case they directly follow the assignment, I don't understand the point (yes, someone can change the assignment, but they can change any other line of code and ruin it - we don't put an assert after every statement ;-)).

Reply:
removed (again)